### ISSUE TITLE
Complete showcase

### ISSUE THREAD

**Issue by uosis** [uosis/laminar-web-components#5]
Use all components

**Comment by mathieuleclaire** [comment]
Hi,
I'd like to help to complete the showcase, but I am not sure to understand how the lib works. I tried to add a simple List, inspired by the TabBar example:

```scala
 List(
        _.slots.default(
          List.ListItem(_.value :=  "first"),
          List.ListItem(_.value := "second")
        )
      )
```
but it only produces this html and nothing appears in the browser:
```html
<mwc-list>
   <mwc-list-item slot=""></mwc-list-item>
   <mwc-list-item slot=""></mwc-list-item>
</mwc-list>
```

Any idea ?

**Comment by raquo** [comment]
You're rendering empty elements, that's why they don't show up in the browser. According to the [documentation](https://github.com/material-components/material-components-web-components/tree/master/packages/list) for this web component you need to put the name of the item you want to display as a child of that item, e.g. "<mwc-list-item>Item 0</mwc-list-item>". In this library this would be done like this: `List.ListItem(_ => "Item 0")`.

The [Laminar intro video](https://www.youtube.com/watch?v=L_AHCkl6L-Q&t=4132s) talks about using web components at the end. This library follows a similar pattern so the video should help you understand how to express things in Laminar that you see in the documentation for a particular web component.

**Comment by mathieuleclaire** [comment]
Unfortunately, it leads to very close result.
```scala
List(
        _=> activatable := true,
        _.slots.default(
          List.ListItem(_=>  "first"),
          List.ListItem(_=> "second"),
        )
      )
```
leads to this:
![Capture d’écran du 2020-12-01 23-01-55](https://user-images.githubusercontent.com/800035/100802143-445a3c80-3429-11eb-8fc4-5c6e46a18c4f.png)

**Comment by raquo** [comment]
Hm, not sure why the styles aren't being applied. The only suspect thing here is `_.slots.default(`, this component's documentation does not mention anything about slots. So maybe this should work:

```
List(
  _=> activatable := true,
  List.ListItem(_ => "first"),
  List.ListItem(_ => "second")
)
```

If it produces the same HTML as the web component's docs / demos show, it really should behave the same as the styles of web components are self-contained as far as I understand.

**Comment by mathieuleclaire** [comment]
It has the same effect. I try this too with the same effect:

```scala
  List(
        _ => activatable := true,
        _ => List.ListItem(_ => "first"),
        _ => List.ListItem(_ => "second"),
      )
```
The fact is that, it produces a ```<mwc-list>``` tag instead of prducing a ```<li class="mwc-list">``` tag. Idem for the ListItems (no ```<ul>``` tags are produced.

**Comment by raquo** [comment]
> The fact is that, it produces a `<mwc-list>` tag instead of prducing a `<li class="mwc-list">` tag. Idem for the ListItems (no `<ul>` tags are produced.

That's ok and is to be expected. Those web components don't produce `<ul>` or `<li>` tags, as you can see on their [demo page](https://material-components.github.io/material-components-web-components/demos/list/). When used correctly they produce HTML like:

```html
    <mwc-list>
      <mwc-list-item>first</mwc-list-item>
      <mwc-list-item>second</mwc-list-item>
    </mwc-list>
```

---

Now that I think of it, It kinda looks like your problem might be caused by you not importing the required JS dependency. For example, in my laminar-examples repo I have a small interface ([Button.scala](https://github.com/raquo/laminar-examples/blob/master/src/main/scala/webcomponents/material/Button.scala)) for mwc-button, it includes this line:

```scala
@js.native
@JSImport("@material/mwc-button", JSImport.Default)
```

I don't see a similar `JSImport `for `mwc-list` or `mwc-list-item` anywhere in this project. Do you have it locally? If not, this is why you're not getting the web component. This project is WIP so I don't know, maybe that's one of the missing things. Did you run `amm generator/main.sc` as mentioned in the readme? Maybe that will generate the right code.

Note: make sure to use JSImport properly, similarly to how it's done in my `Button.scala` example. Keep in mind that Scala uses dead code elimination, and `object`-s are lazy. The video mentions that briefly IIRC. Overall I suggest that you look at my `Button.scala` example because I know that it does work – you can check it out and see it for yourself. And then look at what's missing from the List and ListItem that you're looking at. JsImport is one thing but there could be more issues.

For the sake of completeness – when you add the dependency with JSImport, you will probably get an error about Scala.js being unable to find that dependency. This is to be expected, as you also need to add this dependency to your `npmDependencies` in build.sbt and install it (see Scala.js docs for that).

**Comment by mathieuleclaire** [comment]
There is the js.native and JSImport stuff here in [material.scala](https://github.com/uosis/laminar-web-components/blob/master/material/src/main/scala/material.scala#L2648:L2657)

I tried to regenerate the material from the generator, moving to version 0.19.1, but it still the same result. I am going to investigate other components like Button.

Could it be again an asynchronousity issue ?

**Comment by raquo** [comment]
Ah, funny that github search doesn't quite find that, but yes, JSImport appears to be in place.

Well, I assume you're not getting any errors in the browser console, so that wouIdn't be another asynchrony issue. I'm out of ideas, sorry.

**Comment by uosis** [comment]
@mathieuleclaire if you are still playing around with this, I pushed a fix [here](https://github.com/uosis/laminar-web-components/compare/fix-mwc-list-item).

The problem was that for this particular component, NPM package name does not match the required import, so the list item component was not getting imported. My generator doesn't even have support for different npm package vs import, so at this point if you run the generator, you need to edit the output manually to remove unnecessary package from build.sbt.

As you can see, this whole thing is very messy, as is typical in Javascript world - there are no machine readable definitions for any of these web components, so we are stuck parsing READMEs and then manually tweaking things. I wouldn't be surprised if there are more issues like this here.

**Comment by mathieuleclaire** [comment]
Hi uosis,
Thanks for your answer. I tried your fix but it does not seem to display properly a list:
![image](https://user-images.githubusercontent.com/800035/102182082-f70baf80-3eab-11eb-8c78-9c78e992eba5.png)

**Comment by uosis** [comment]
This looks like the js import is still missing - did you rebuild components package from that branch?

```shell
cd material
sbt publishLocal

cd ../examples/showcase
sbt fastOptJS::webpack
```

**Comment by mathieuleclaire** [comment]
This works ! Thanks.
Could you merge this fix in master ?

![image](https://user-images.githubusercontent.com/800035/103523969-111e2980-4e7d-11eb-990d-6d5d23d32fcb.png)

**Comment by uosis** [comment]
> My generator doesn't even have support for different npm package vs import, so at this point if you run the generator, you need to edit the output manually to remove unnecessary package from build.sbt.

I just manually edited generated output to make this work - I don't really want to merge it to master because it will get overwritten unless generator is updated to support this.

**Comment by mathieuleclaire** [comment]
OK, so let's fix the generator then. I don't know how it works, but is it enough to switch "list" to "list-item" [here](https://github.com/uosis/laminar-web-components/blob/master/generator/material.sc#L1310) ?

**Comment by uosis** [comment]
No, it's not as simple - we need to add support to the generator and component definitions to specify `@JSImport` name independently from NPM package name. Currently it assumes that they will always be the same.

So you would have something like this in definitions:

```scala
npmPackage = NpmPackage(name = "@material/mwc-list-item", version = "0.18.0"),
jsImport = JSImport(name = "@material/mwc-list/mwc-list-item"),
```

and the generator would need to use that to emit proper imports and dependencies.

It's not a huge amount of work, but it does require updating all definitions (or perhaps having jsImport default to npmPackage name).

**Comment by mathieuleclaire** [comment]
Hi,
I modified the npmPackage name in the generator, and it works fine: https://github.com/uosis/laminar-web-components/pull/9
